Modules/_abc.c

```diff
@@ -0,0 +1,1049 @@
+    if (result < 0) {
+        return NULL;
+    }
+    if (_add_to_registry(self, subclass) < 0) {
+        return NULL;
+    }
+    Py_INCREF(subclass);
+    /* Invalidate negative cache */
+    Py_DECREF(abc_invalidation_counter);
+    one = PyLong_FromLong(1);
+    abc_invalidation_counter = PyNumber_Add(abc_invalidation_counter, one);
+    Py_DECREF(one);
+    return subclass;
+}
+
+PyDoc_STRVAR(_abc_instancecheck_doc,
+"Internal ABC helper for instance checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_instancecheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *result, *subclass, *subtype, *instance = NULL;
+    int incache;
+    if (!PyArg_UnpackTuple(args, "_abc_instancecheck", 2, 2, &self, &instance)) {
+        return NULL;
+    }
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
+    /* Inline the cache checking. */
+    incache = _in_cache(self, subclass);
+    if (incache < 0) {
+        Py_DECREF(subclass);
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_DECREF(subclass);
+        Py_RETURN_TRUE;
+    }
+    subtype = (PyObject *)Py_TYPE(instance);
+    if (subtype == subclass) {
+        incache = _in_negative_cache(self, subclass);
+        if (incache < 0) {
+            Py_DECREF(subclass);
+            return NULL;
+        }
+        if ((PyObject_RichCompareBool(_get_negative_cache_version(self),
+                                      abc_invalidation_counter, Py_EQ) > 0) && incache) {
+            Py_DECREF(subclass);
+            Py_RETURN_FALSE;
+        }
+        /* Fall back to the subclass check. */
+        result = PyObject_CallMethod(self, "__subclasscheck__", "O", subclass);
+        Py_DECREF(subclass);
+        return result;
+    }
+    result = PyObject_CallMethod(self, "__subclasscheck__", "O", subclass);
+    if (!result) {
+        Py_DECREF(subclass);
+        return NULL;
+    }
+    if (result == Py_True) {
+        Py_DECREF(subclass);
+        return Py_True;
+    }
+    Py_DECREF(result);
+    Py_DECREF(subclass);
+    return PyObject_CallMethod(self, "__subclasscheck__", "O", subtype);
+}
+
+
+PyDoc_STRVAR(_abc_subclasscheck_doc,
+"Internal ABC helper for subclasss checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_subclasscheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *subclasses, *subclass = NULL;
+    PyObject *ok, *mro, *key, *rkey;
+    Py_ssize_t pos;
+    int incache, result;
+    if (!PyArg_UnpackTuple(args, "_abc_subclasscheck", 2, 2, &self, &subclass)) {
+        return NULL;
+    }
+
+    /* 1. Check cache. */
+    incache = _in_cache(self, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_RETURN_TRUE;
+    }
+    /* 2. Check negative cache; may have to invalidate. */
+    incache = _in_negative_cache(self, subclass);
+    if (incache < 0) {
+        return NULL;
+    }
+    if (PyObject_RichCompareBool(_get_negative_cache_version(self),
```

Comment:
Handle case when `PyObject_RichCompareBool` returns -1.

Reply:
And check the result of `_get_negative_cache_version` call.

Reply:
The ` _get_negative_cache_version` should never fail. I will refactored it, so that it is clear.